### PR TITLE
Add addon/network/aad support for AKS

### PR DESCRIPTION
**Pull request by yuwzho**
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

azure_rm_aks

##### ADDITIONAL INFORMATION

2.7.2

```paste below

```

### CODE REVIEW

**Comment by yuwzho** [comment]
@Fred-sun Please help me to test the [AKS integration test](https://github.com/VSChina/ansible/blob/yuwzho-aks/test/integration/targets/azure_rm_aks/tasks/main.yml) since this is disabled in the CI.

**Comment by Fred-sun** [comment]
@yuwzho I have tested, meet error, I will retest after you update. Error info as below:
```e.g.
TASK [include_tasks] ********************************************************************************************
fatal: [localhost]: FAILED! => {"reason": "A malformed block was encountered while loading a block"}
	to retry, use: --limit @/home/fred/task/aks/task.retry

PLAY RECAP ******************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=1    skipped=0

**Comment by yuwzho** [comment]
@Fred-sun Seems your error has none business with the  aks module itself, but something wrong with your playbook

**Comment by Fred-sun** [comment]
@yuwzho Thanks for your advice. There's one line with the wrong indent,  I meet another error, error info as below:
```e.g.
TASK [Create an AKS instance] ***********************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: type object 'ContainerServiceClient' has no attribute 'models'
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/fred/.ansible/tmp/ansible-tmp-1547778552.0-163331998703078/AnsiballZ_azure_rm_aks.py\", line 113, in <module>\n    _ansiballz_main()\n  File \"/home/fred/.ansible/tmp/ansible-tmp-1547778552.0-163331998703078/AnsiballZ_azure_rm_aks.py\", line 105, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/fred/.ansible/tmp/ansible-tmp-1547778552.0-163331998703078/AnsiballZ_azure_rm_aks.py\", line 48, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_azure_rm_aks_payload_6kMWta/__main__.py\", line 723, in <module>\n  File \"/tmp/ansible_azure_rm_aks_payload_6kMWta/__main__.py\", line 719, in main\n  File \"/tmp/ansible_azure_rm_aks_payload_6kMWta/__main__.py\", line 430, in __init__\n  File \"/tmp/ansible_azure_rm_aks_payload_6kMWta/ansible_azure_rm_aks_payload.zip/ansible/module_utils/azure_rm_common.py\", line 310, in __init__\n  File \"/tmp/ansible_azure_rm_aks_payload_6kMWta/__main__.py\", line 550, in exec_module\n  File \"/tmp/ansible_azure_rm_aks_payload_6kMWta/__main__.py\", line 586, in create_update_aks\n  File \"/tmp/ansible_azure_rm_aks_payload_6kMWta/__main__.py\", line 673, in create_agent_pool_profile_instance\n  File \"/tmp/ansible_azure_rm_aks_payload_6kMWta/ansible_azure_rm_aks_payload.zip/ansible/module_utils/azure_rm_common.py\", line 864, in containerservice_models\nAttributeError: type object 'ContainerServiceClient' has no attribute 'models'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
	to retry, use: --limit @/home/fred/task/aks/1.retry

PLAY RECAP ******************************************************************************************************
localhost                  : ok=5    changed=1    unreachable=0    failed=1    skipped=0

**Comment by yuwzho** [comment]
@Fred-sun Please update the package to `azure-mgmt-containerservice==4.4.0`

**Comment by Fred-sun** [comment]
@yuwzho Yes, This error caused by azure-mgmt-containerservice, After update, test successful. Continue with other tests!

**Comment by Fred-sun** [comment]
@yungezz @zikalino

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py [hunk 1]
```diff
@@ -96,6 +96,57 @@
...
+                description: The server AAD application secret.
+            tenant_id:
+                description:
+                    - The AAD tenant ID to use for authentication.
+                    - If not specified, will use the tenant of the deployment subscription.
+        version_added: 2.8
+    addon:
+        description:
+            - Profile of managed cluster add-on.
+            - Key can be C(http_application_routing), C(monitoring), C(virtual_node).
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_aks.py:152]: make the 3 choices?

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py [hunk 2]
```diff
@@ -597,9 +715,55 @@ def get_aks_kubeconfig(self):
 
         :return: AKS instance kubeconfig
         '''
-        access_profile = self.containerservice_client.managed_clusters.get_access_profiles(
-            self.resource_group, self.name, "clusterUser")
-        return base64.b64decode(access_profile.kube_config)
+        access_profile = self.containerservice_client.managed_clusters.get_access_profile(self.resource_group, self.name, "clusterUser")
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_aks.py]: specify keyword arguments

**Comment by Fred-sun** [comment]
@yungezz @zikalino Please help review and push for merged! Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py [hunk 3]
```diff
@@ -329,6 +406,24 @@ def create_agent_pool_profiles_dict(agentpoolprofiles):
...
+    service_cidr=dict(type='str'),
+    dns_service_ip=dict(type='str'),
+    docker_bridge_cidr=dict(type='str')
+)
+
+
+aad_profile_spec = dict(
+    client_app_id=dict(type='str'),
+    server_app_id=dict(type='str'),
+    server_app_secret=dict(type='str'),
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_aks.py]: no_log

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py [hunk 4]
```diff
@@ -419,22 +530,20 @@ def exec_module(self, **kwargs):
             # For now Agent Pool cannot be more than 1, just remove this part in the future if it change
             agentpoolcount = len(self.agent_pool_profiles)
             if agentpoolcount > 1:
-                self.fail(
-                    'You cannot specify more than one agent_pool_profiles currently')
+                self.fail('You cannot specify more than one agent_pool_profiles currently')
 
-            if response:
-                self.results = response
+            available_versions = self.get_all_versions()
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_aks.py:542]: how about add getting available version in facts module?

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py [hunk 5]
```diff
@@ -96,6 +96,92 @@
...
+        suboptions:
+            network_plugin:
+                description:
+                    - Network plugin used for building Kubernetes network.
+                    - This property cannot been changed.
+                choices:
+                    - azure
+                    - kubenet
+            network_policy:
+                description: Network policy used for building Kubernetes network.
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_aks.py:114]: so what should that policy be exactly?

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py [hunk 6]
```diff
@@ -96,6 +96,92 @@
...
+            - Profile of managed cluster add-on.
+            - Key can be C(http_application_routing), C(monitoring), C(virtual_node).
+            - Value must be a dict contains a bool variable C(enabled).
+        type: dict
+        suboptions:
+            http_application_routing:
+                description:
+                    - The HTTP application routing solution makes it easy to access applications that are deployed to your cluster.
+                type: dict
+                suboptions:
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_aks.py:160]: do we need suboptions if it's jut one bool?

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py [hunk 7]
```diff
@@ -96,6 +96,92 @@
...
+                    - You don't need to wait for Kubernetes cluster autoscaler to deploy VM compute nodes to run the additional pods.
+                type: dict
+                suboptions:
+                    enabled:
+                        description:
+                            - Whether the solution enabled.
+                        type: bool
+                    subnet_resource_id:
+                        description:
+                            - Subnet associdated to the cluster.
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_aks.py:190]: shall we support dictionary here?

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py (hunk 6 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_aks.py:160]: I wanna keep consistency for all addons' parameters. Although http_application_routing has only one enabled parameter, other addons contain more than one.

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py (hunk 5 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_aks.py:114]: which can be used to set  the network rule of pod: https://kubernetes.io/docs/concepts/services-networking/network-policies/

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_aks.py:114]: ok, so the most likely scenario will be that user will use lookup to load it from file or somewhere else, right?

File: lib/ansible/modules/cloud/azure/azure_rm_aks_facts.py [hunk 8]
```diff
@@ -33,6 +33,19 @@
     tags:
         description:
             - Limit results by providing a list of tags. Format tags as 'key' or 'key:value'.
+    show_kubeconfig:
+        description:
+            - Show kubeconfig of the AKS cluster.
+            - Note the operation will cost more network overhead, not recommond to use when listing AKS.
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_aks_facts.py]: recommond -> recommend -> not recommended when listing multiple instances

File: lib/ansible/modules/cloud/azure/azure_rm_aks_facts.py [hunk 9]
```diff
@@ -33,6 +33,19 @@
...
+        version_added: 2.8
+        choices:
+            - user
+            - admin
+    show_available_versions_in_location:
+        description:
+            - Show available kubernetes versions in a specific location.
+            - Note, When this field set, nothing will be queried in your subscription but only available kubernetes versions.
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_aks_facts.py]: i don't quite understand what it means.
does it mean no instances will be listed?
if so then perhaps this should be in a separate module?

**Comment by yuwzho** [comment]
ready_for_review

File: lib/ansible/modules/cloud/azure/azure_rm_aks.py (hunk 5 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_aks.py:114]: add doc link for user's reference?

File: lib/ansible/modules/cloud/azure/azure_rm_aks_version.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2018 Yuwei Zhou, <yuwzho@microsoft.com>
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_aks_version.py:3]: 2019

**Review by yungezz — APPROVED** [review]
minor comments.

**Comment by yungezz** [comment]
shipit

**Review by zikalino — APPROVED** [review]

**Pull request merged**